File: src/webmention/receive.js

const got = require('got')
const config = require('./../config')
const fsp = require('fs').promises
const md5 = require('md5')
const { mf2 } = require("microformats-parser");
const dayjs = require('dayjs')
const utc = require('dayjs/plugin/utc')
dayjs.extend(utc)

function isValidUrl(url) {
	return url !== undefined &&
		(url.startsWith("http://") || url.startsWith("https://"))
}

function isValidDomain(url) {
	return config.allowedWebmentionSources.some(domain => {
		return url.indexOf(domain) !== -1
	})
}

/**
Remember, TARGET is own domain, SOURCE is the article to process
 https://www.w3.org/TR/webmention/#sender-notifies-receiver
 example:
		POST /webmention-endpoint HTTP/1.1
		Host: aaronpk.example
		Content-Type: application/x-www-form-urlencoded

		source=https://waterpigs.example/post-by-barnaby&
		target=https://aaronpk.example/post-by-aaron


		HTTP/1.1 202 Accepted
*/	 
function validate(request) {
	return request.type === "application/x-www-form-urlencoded" &&
		request.body !== undefined &&
		isValidUrl(request?.body?.source) &&
		isValidUrl(request?.body?.target) &&
		request?.body?.source !== request?.body?.target &&
		isValidDomain(request?.body?.target)
}

async function isValidTargetUrl(target) {
	try {
		await got(target)
		return true
	} catch(unknownTarget) {
	}
	console.log(` ABORT: invalid target url`)
	return false
}

function asPath(source, target) {
	const filename = md5(`source=${source},target=${target}`)
	const domain = config.allowedWebmentionSources.find(d => target.indexOf(d) >= 0)
	return `data/${domain}/${filename}.json`
}

async function deletePossibleOlderWebmention(source, target) {
	await fsp.unlink(asPath(source, target))
}

async function saveWebmentionToDisk(source, target, mentiondata) {
	await fsp.writeFile(asPath(source, target), mentiondata, 'utf-8')
}

function publishedNow() {
	return dayjs.utc().utcOffset(config.utcOffset).format("YYYY-MM-DDTHH:mm:ss")
}

function parseBodyAsIndiewebSite(source, target, hEntry) {
	function shorten(txt) {
		if(!txt || txt.length <= 250) return txt
		return txt.substring(0, 250) + "..."
	}

	const name = hEntry.properties?.name?.[0]
	const authorPropName = hEntry.properties?.author?.[0]?.properties?.name?.[0]
	const authorValue = hEntry.properties?.author?.[0]?.value
	const picture = hEntry.properties?.author?.[0]?.properties?.photo?.[0]
	const summary = hEntry.properties?.summary?.[0]
	const contentEntry = hEntry.properties?.content?.[0]?.value
	const bridgyTwitterContent = hEntry.properties?.["bridgy-twitter-content"]?.[0]
	const publishedDate = hEntry.properties?.published?.[0]
	const uid = hEntry.properties?.uid?.[0]
	const url = hEntry.properties?.url?.[0]
	const type = hEntry.properties?.["like-of"]?.length ? "like" : (hEntry.properties?.["bookmark-of"]?.length ? "bookmark" : "mention" )

	return {
		author: {
			name: authorPropName ? authorPropName : authorValue,
			picture: picture?.value ? picture?.value : picture
		},
		name: name,
		content: bridgyTwitterContent ? shorten(bridgyTwitterContent) : (summary ? shorten(summary) : shorten(contentEntry)),
		published: publishedDate ? publishedDate : publishedNow(),
		type,
		// Mastodon uids start with "tag:server", but we do want indieweb uids from other sources 
		url: uid && uid.startsWith("http") ? uid : (url ? url : source),
		source,
		target
	}
}

function parseBodyAsNonIndiewebSite(source, target, body) {
	const title = body.match(/<title>(.*?)<\/title>/)?.splice(1, 1)[0]

	return {
		author: {
			name: source
		},
		name: title,
		content: title,
		published: publishedNow(),
		url: source,
		type: "mention",
		source,
		target
	}
}

async function processSourceBody(body, source, target) {
	if(body.indexOf(target) === -1) {
		console.log(` ABORT: no mention of ${target} found in html src of source`)
		return
	}

	// fiddle: https://aimee-gm.github.io/microformats-parser/
	const microformat = mf2(body, {
		// WHY? crashes on relative URL, should be injected using Jest. Don't care. 
		baseUrl: source.startsWith("http") ? source : `http://localhost/${source}`
	})
	const hEntry = microformat.items.filter(itm => itm?.type?.includes("h-entry"))?.[0]

	const data = hEntry ? parseBodyAsIndiewebSite(source, target, hEntry) : parseBodyAsNonIndiewebSite(source, target, body)
	await saveWebmentionToDisk(source, target, JSON.stringify(data))
	console.log(` OK: webmention processed`)
}

async function receive(body) {
	if(!isValidTargetUrl(body.target)) return

	let src = { body: "" }
	try {
		src = await got(body.source)
	} catch(unknownSource) {
		console.log(` ABORT: invalid source url: ` + unknownSource)
		await deletePossibleOlderWebmention(body.source, body.target)
		return
	}
	await processSourceBody(src.body, body.source, body.target)
} 

module.exports = {
	receive,
	validate
}

File: app/webmention/receive.go

package webmention

import (
	"crypto/md5"
	"encoding/json"
	"fmt"
	"github.com/wgroeneveld/go-jamming/common"
	"github.com/wgroeneveld/go-jamming/rest"
	"io/fs"
	"io/ioutil"
	"net/url"
	"os"
	"regexp"
	"strings"

	"github.com/rs/zerolog/log"
	"willnorris.com/go/microformats"
)

type Mention struct {
	Source string
	Target string
}

func (wm *Mention) String() string {
    return fmt.Sprintf("source: %s, target: %s", wm.Source, wm.Target)
}

func (wm *Mention) asPath(conf *common.Config) string {
	filename := fmt.Sprintf("%x", md5.Sum([]byte("source=" + wm.Source+ ",target=" + wm.Target)))
	domain, _ := conf.FetchDomain(wm.Target)
	return conf.DataPath + "/" + domain + "/" + filename + ".json"
}

func (wm *Mention) sourceUrl() *url.URL {
	url, _ := url.Parse(wm.Source)
	return url
}

// used as a "class" to iject dependencies, just to be able to test. Do NOT like htis. 
// Is there a better way? e.g. in validate, I just pass rest.Client as an arg. Not great either. 
type Receiver struct {
	RestClient rest.Client
	Conf       *common.Config
}

func (recv *Receiver) Receive(wm Mention) {
	log.Info().Str("Webmention", wm.String()).Msg("OK: looks valid")
	body, geterr := recv.RestClient.GetBody(wm.Source)

	if geterr != nil {
        log.Warn().Str("source", wm.Source).Msg("  ABORT: invalid url")
		recv.deletePossibleOlderWebmention(wm)
		return
	}

	recv.processSourceBody(body, wm)
}

func (recv *Receiver) deletePossibleOlderWebmention(wm Mention) {
	os.Remove(wm.asPath(recv.Conf))
}

func getHEntry(data *microformats.Data) *microformats.Microformat {
	for _, itm := range data.Items {
		if common.Includes(itm.Type, "h-entry") {
			return itm
		}
	}
	return nil
}


func (recv *Receiver) processSourceBody(body string, wm Mention) {
	if !strings.Contains(body, wm.Target) {
		log.Warn().Str("target", wm.Target).Msg("ABORT: no mention of target found in html src of source!")
		return
	}

	data := microformats.Parse(strings.NewReader(body), wm.sourceUrl())
	indieweb := recv.convertBodyToIndiewebData(body, wm, getHEntry(data))

	recv.saveWebmentionToDisk(wm, indieweb)
	log.Info().Str("file", wm.asPath(recv.Conf)).Msg("OK: Webmention processed.")
}

func (recv *Receiver) convertBodyToIndiewebData(body string, wm Mention, hEntry *microformats.Microformat) *indiewebData {
	if hEntry == nil {
		return recv.parseBodyAsNonIndiewebSite(body, wm)
	}
	return recv.parseBodyAsIndiewebSite(hEntry, wm)
}

func (recv *Receiver) saveWebmentionToDisk(wm Mention, indieweb *indiewebData) {
	jsonData, jsonErr := json.Marshal(indieweb)
	if jsonErr != nil {
		log.Err(jsonErr).Msg("Unable to serialize Webmention into JSON")
	}
	err := ioutil.WriteFile(wm.asPath(recv.Conf), jsonData, fs.ModePerm)
	if err != nil {
		log.Err(err).Msg("Unable to save Webmention to disk")
	}
}

// TODO I'm smelling very unstable code, apply https://golang.org/doc/effective_go#recover here?
// see https://github.com/willnorris/microformats/blob/main/microformats.go
func (recv *Receiver) parseBodyAsIndiewebSite(hEntry *microformats.Microformat, wm Mention) *indiewebData {
	name := mfStr(hEntry, "name")
	pic := mfStr(mfProp(hEntry, "author"), "photo")
	mfType := determineMfType(hEntry)

	return &indiewebData{
		Name: name,
		Author: indiewebAuthor{
			Name: determineAuthorName(hEntry),
			Picture: pic,
		},
		Content: determineContent(hEntry),
		Url: determineUrl(hEntry, wm.Source),
		Published: determinePublishedDate(hEntry, recv.Conf.UtcOffset),
		Source: wm.Source,
		Target: wm.Target,
		IndiewebType: mfType,
	}
}

func (recv *Receiver) parseBodyAsNonIndiewebSite(body string, wm Mention) *indiewebData {
	r := regexp.MustCompile(`<title>(.*?)<\/title>`)
	titleMatch := r.FindStringSubmatch(body)
	title := wm.Source
	if titleMatch != nil {
		title = titleMatch[1]
	}
	return &indiewebData{
		Author: indiewebAuthor{
			Name: wm.Source,
		},
		Name: title,
		Content: title,
		Published: publishedNow(recv.Conf.UtcOffset),
		Url: wm.Source,
		IndiewebType: "mention",
		Source: wm.Source,
		Target: wm.Target,
	}
}
